### PR TITLE
feature: database transaction logging

### LINKED ISSUE #1480: Feature Request: Create Transaction Log Collection for Database
**Is your feature request related to a problem? Please describe.**
We don't track when a record is created or updated in the database.

**Describe the solution you'd like**
Setting up a new collection dedicated to logging all transactions or changes made within the database. This collection will serve as a detailed history of all data interactions, providing an extensive audit trail.

### PR DESCRIPTION
**Pull request by LakshyaSatpal**
**What kind of change does this PR introduce?**

**Issue Number:**

Fixes #1480 

**Snapshots/Videos:**
The transaction file will look something like this:

image
first 3 logs are when a user logs in, and last log is when an organization is created. 

every log will have info about:
- timestamp (compulsory)
- model (compulsory)
- type (compulsory)
- update (optional, the updated fields in the model)
- query (optional, the document that was queried)

For type=CREATE transactions, there won't be any `update` or `query` logged.
For type=DELETE, `query` will be logged

**If relevant, did you update the documentation?**

**Summary**
Created a `src/libraries/dbLogger.ts` file, and a new middleware function that registers logging middleware for a mongoose model. 
(hence, we don't need to log transactions manually in resolver files)

Currently, this middleware is added only in `User` model. 

**Does this PR introduce a breaking change?**
No

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

### CODE REVIEW

**Comment by LakshyaSatpal** [comment]
@EshaanAgg Could you please review this?

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -126,6 +126,12 @@ jobs:
 
       - name: Generate Refresh Token Secret
         run: echo "REFRESH_TOKEN_SECRET=$(openssl rand -hex 32)" >> $GITHUB_ENV
+      
+      - name: Set up Transaction Logs path
+        run: |
+          mkdir -p logs
+          touch logs/transaction.log
+          echo "TRANSACTION_LOG_PATH=${GITHUB_WORKSPACE}/logs/transaction.log" >> $GITHUB_ENV
```
Comment (palisadoes) [inline .github/workflows/pull-request.yml:134]: 1. Make the transaction log location a part of the `.env` file configuration
2. Update the `.env.sample` file with this information
3. Update `setup.ts` accordingly

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. Make the transaction log location a part of the `.env` file configuration
2. Update the `.env.sample` file with this information
3. Update `setup.ts` accordingly

File: src/models/User.ts [hunk 2]
```diff
@@ -283,6 +285,8 @@ userSchema.plugin(mongoosePaginate);
 const userModel = (): PaginateModel<InterfaceUser> =>
   model<InterfaceUser, PaginateModel<InterfaceUser>>("User", userSchema);
 
+createLoggingMiddleware(userSchema, "User");
```
Comment (EshaanAgg) [inline src/models/User.ts:288]: This is so elegant! Thanks for putting in the work to research the same!

**Review by EshaanAgg — CHANGES_REQUESTED** [review]

File: src/models/User.ts (hunk 2 above)
Comment (LakshyaSatpal) [reply src/models/User.ts:288]: Thank you @EshaanAgg

Comment (LakshyaSatpal) [reply src/models/User.ts:288]: Should I add this middleware in all model files?

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. Fix the failing tests
2. Update `setup.ts` to prompt for the parameters. The installation user experience must be flawless. We have been getting kudos for the ease of operation. We cannot afford for that reputation to be tarnished

**Comment by LakshyaSatpal** [comment]
@palisadoes The tests are passing now.
I've tried to maintain user-friendliness in `setup.ts`. Let me explain the flow:
 - by default, `.env.sample` will have 2 params `LOG=false` and `LOG_PATH=./logs/transaction.log`, as suggested by @EshaanAgg 
 - in setup, user will get a prompt `Would you like to enable logging for the database transactions?`. 
        -  if yes, it will ask for `Would you like to provide a custom path for storing logs?`. 
                - If yes, it will ask for a readable and writable file, and `LOG=true` and `LOG_PATH=<absolute-path-of-file>`. 
                - if no, `LOG=true` and `LOG_PATH=<absolute-path-of-logs/transaction.log>`
        - if no, it will skip everything, and move to next step in `setup.ts`.
 
 Is there any specific prompt that we can add here?

File: src/libraries/dbLogger.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
+import { LOG, LOG_PATH, TransactionLogTypes } from "../constants";
+import type { Query, Schema, Document } from "mongoose";
+import winston from "winston";
+
+export type TransactionLogInfo = {
+  timestamp: string;
```
Comment (palisadoes) [inline src/libraries/dbLogger.ts:6]: 1. There are no tests written for this file. 
2. Try to get the test coverage up to 100%

File: .env.sample [hunk 4]
```diff
@@ -49,6 +49,12 @@ SMTP_USERNAME=
 SMTP_PORT=
 SMTP_SSL_TLS=
 
+# Enable or disable the storage of logs
+LOG=false
+
+
+# Path of file that will store logs
+LOG_PATH=./logs/transaction.log
```
Comment (palisadoes) [inline .env.sample:57]: Update the `.gitignore` file. Add:

1. The `logs/` directory and all files in it. There are examples for the `images/` directory
2. All files ending in `.log` 

We don't wan these being added to the repo

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comments

File: .env.sample (hunk 4 above)
Comment (LakshyaSatpal) [reply .env.sample:57]: I added a `logs/README.md` file as suggested by @EshaanAgg. I've changed the `.gitignore` file to ignore all files inside the `logs` dir except README.md.
And, ignoring all the `.log` files.

File: src/libraries/dbLogger.ts (hunk 3 above)
Comment (LakshyaSatpal) [reply src/libraries/dbLogger.ts:6]: done

File: src/models/User.ts (hunk 2 above)
Comment (EshaanAgg) [reply src/models/User.ts:289]: No. Let's merge this and wait a few days so all the developers can try it. If we get a positive feedback, and there were no unexpected errors encountered, then we can work on adding the same for all models.

**Review by EshaanAgg — APPROVED** [review]
LGTM. Thanks.

**Review by palisadoes — APPROVED** [review]

**Pull request merged**